### ISSUE TITLE
`storeDateTimesAsText` dialect aware

### ISSUE THREAD

**Issue by davidmartos96** [simolus3/moor#2949]
Hello!
Do you think the `storeDateTimesAsText` option should only have effect when using the sqlite dialect? Otherwise it can break when generating queries in Postgres and comparing dates, because the `JULIANDAY` function is sent to Postgres.

Maybe it makes sense to have something similar to this in the `SqlTypes` class, so that places using it only do it when the dialect is sqlite. Maybe renaming the field could be nice as well.

```dart
final bool _storeDateTimesAsText;

bool get storeDateTimesAsText => dialect == SqlDialect.sqlite && _storeDateTimesAsText;
```

https://github.com/simolus3/drift/blob/74fb2696159904e91ad8a98335b1bb9362363b2b/drift/lib/src/runtime/query_builder/expressions/expression.dart#L442

**Comment by simolus3** [comment]
Pretty much everything that touches the datetime type in drift is tightly coupled to sqlite3 at the moment. Most of the methods drift provides for datetime modification also won't work with postgres. It would be nice if implementations could provide a way to intercept default types somehow (so that `dateTime()` creates a proper datetime column on postgres).

At the moment, my recommendation is to avoid that type entirely if you're targeting other databases (and use a dialect-specific type instead).
I can change the implementation so that the option only applies to sqlite3, but that doesn't change the fact that the default type for date times is broken on postgres.

**Comment by davidmartos96** [comment]
@simolus3 Yes, I totally get that. The problem came from an user of `electricsql` which was reusing the generated drift schema on Postgres and had that option enabled. And unfortunately the error was difficult to debug.

One thing to note is that this issue, related to how the comparator expression gets generated, doesn't occur in latest git version of drift, thanks to the fix on Variable's with CustomTypes, but it hasn't been released yet.

**Comment by simolus3** [comment]
The fix was part of drift 2.17. If there are additional problems where custom types don't propagate through the expression builder API, I'll fix them as well. I think that solves this issue (until we eventually have proper support for datetimes across dialects).

**Comment by davidmartos96** [comment]
@simolus3 Yes, with the fix for the CustomTypes is less of deal breaker.
Thank you!
